cmd/snap-confine/mount-support.c

```diff
@@ -357,10 +358,16 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 		// assumes to find snap-exec
 		sc_must_snprintf(dst, sizeof dst, "%s/usr/lib/snapd",
 				 scratch_dir);
-		// FIXME: take re-exec into account, i.e. this may not
-		//        actually be the path we need, we need snapd
-		//        to tell us what path to use instead.
-		const char *src = LIBEXECDIR;
+
+		// bind mount the current $ROOT/usr/lib/snapd path,
+		// where $ROOT is either "/" or the "/snap/core/current"
+		// that we are re-execing from
+		char *src = LIBEXECDIR;
+		char self[PATH_MAX + 1] = { 0, };
+		if (readlink("/proc/self/exe", self, sizeof self) > 0) {
```

Comment:
Can we use `sizeof(self) - 1` here? readlink() doesn't append a NULL byte so if for some reason the kernel returned too much, we are safe (since you initialized 'self' to all NULLs). That seems to be your intent for using PATH_MAX + 1, but let's not trust there are no kernel bugs either.

Reply:
I might also point out that the not all /proc/*/exes are symlinks, but they will be for /proc/self/exe for snap-confine (ie, kernel processes won't be a symlink). If we are guarding against kernel bugs, we could:
```
if (strstr(self, "/snap-confine") == NULL) {
    die(...);
}
```

This is a harder kernel bug to imagine happening, but it's a cheap check. @zyga, I didn't think of this one in the 3621 review. Follow-up PR would be fine there.

Reply:
@jdstrand ack, will do.